src/ibmca_ec.c

```diff
@@ -0,0 +1,918 @@
+    if (!ica_privkey) {
+        IBMCAerr(IBMCA_F_ICA_EC_KEY_NEW, IBMCA_R_EC_INTERNAL_ERROR);
+        goto end;
+    }
+
+    /* Get private (D) value from EC_KEY */
+    bn_d = (BIGNUM*)EC_KEY_get0_private_key((EC_KEY*)ecdh);
+    if (bn_d == NULL) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDH_COMPUTE_KEY, IBMCA_R_EC_INTERNAL_ERROR);
+        goto end;
+    }
+
+    /* Format (D) as char array, with leading zeros if necessary */
+    n = privlen - BN_num_bytes(bn_d);
+    for (i = 0; i < n; i++)
+        D[i] = 0x00;
+    BN_bn2bin(bn_d, &(D[n]));
+
+    /* Initialize private ICA_EC_KEY with (D) */
+    rc = p_ica_ec_key_init(NULL, NULL, D, ica_privkey);
+    if (rc != 0) {
+        IBMCAerr(IBMCA_F_ICA_EC_KEY_INIT, rc);
+        goto end;
+    }
+
+    /* Allocate memory for shared secret z, will be freed by caller */
+    if ((z_buf = OPENSSL_malloc(privlen)) == NULL) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDH_COMPUTE_KEY, IBMCA_R_EC_INTERNAL_ERROR);
+        goto end;
+    }
+
+    /* Calculate shared secret z */
+    rc = p_ica_ecdh_derive_secret(ibmca_handle, ica_privkey, ica_pubkey, z_buf,
+                                  privlen);
+    if (rc != 0) {
+        /* Possibly no suitable adapter. */
+        OPENSSL_free(z_buf);
+
+ #ifdef OLDER_OPENSSL
+        goto end;
+ #else
+        EC_KEY_METHOD_get_compute_key(ossl_ec, &compute_key_sw);
+        if (compute_key_sw == NULL) {
+            IBMCAerr(IBMCA_F_ICA_ECDH_DERIVE_SECRET, rc);
+            goto end;
+        }
+
+        ret = compute_key_sw(pout, poutlen, pub_key, ecdh);
+        goto end;
+ #endif
+    }
+
+    *pout = z_buf;
+    *poutlen = privlen;
+
+    ret = 1;
+
+end:
+    p_ica_ec_key_free(ica_pubkey);
+    p_ica_ec_key_free(ica_privkey);
+
+    return ret;
+}
+
+ #ifndef OLDER_OPENSSL
+/**
+ * ECDSA signing method (replaces ossl_ecdsa_sign).
+ *
+ * returns 1 if success
+ *         0 if error
+ */
+int ibmca_ecdsa_sign(int type, const unsigned char *dgst, int dlen,
+             unsigned char *sig_array, unsigned int *siglen,
+             const BIGNUM *kinv, const BIGNUM *r, EC_KEY *eckey)
+{
+    ECDSA_SIG *sig;
+    const BIGNUM *bn_r, *bn_s;
+    const EC_GROUP *group;
+    int n, i, r_len;
+
+    /* Check parms: precomputed (k,r) are not supported by ibmca */
+    if (kinv != NULL || r != NULL) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDSA_SIGN, IBMCA_R_EC_INVALID_PARM);
+        return 0;
+    }
+
+    /* Create signature */
+    sig = ibmca_ecdsa_sign_sig(dgst, dlen, NULL, NULL, eckey);
+    if (sig == NULL) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDSA_SIGN, IBMCA_R_EC_INTERNAL_ERROR);
+        *siglen = 0;
+        return 0;
+    }
+
+    /* Determine r-length */
+    if ((group = EC_KEY_get0_group(eckey)) == NULL) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDSA_SIGN, IBMCA_R_EC_INTERNAL_ERROR);
+        return 0;
+    }
+    r_len = (EC_GROUP_get_degree(group) + 7) / 8;
+
+    /* Get (r,s) from ECDSA_SIG */
+    ECDSA_SIG_get0(sig, &bn_r, &bn_s);
```

Comment:
Is this a void function ? Otherwise some kind of returncode handling would be needed.